core/store/src/db/rocksdb.rs

```diff
@@ -49,14 +49,49 @@ unsafe impl Send for RocksDB {}
 unsafe impl Sync for RocksDB {}
 
 impl RocksDB {
-    /// Opens the database either in read only or in read/write mode depending
-    /// on the `mode` parameter specified in the store_config.
-    pub fn open(path: &Path, store_config: &StoreConfig, mode: Mode) -> io::Result<RocksDB> {
+    /// Opens the database.
+    ///
+    /// `path` specifies location of the database.  It’s assumed that it has
+    /// been resolved based on configuration in `store_config` and thus path
+    /// configuration in `store_config` is ignored.
+    ///
+    /// `store_config` specifies other storage configuration such open files
+    /// limit or whether to enable collection of statistics.
+    ///
+    /// `mode` specifies whether to open the database in read/write or read-only
+    /// mode.  In the latter case, the database will not be created if it
+    /// doesn’t exist nor any migrations will be performed if the database has
+    /// database version different than expected.
+    pub fn open(path: &Path, store_config: &StoreConfig, mode: Mode) -> io::Result<Self> {
+        Self::open_with_columns(path, store_config, mode, DBCol::iter())
+    }
+
+    /// Opens the database with given set of column families configured.
+    ///
+    /// With cold storage, we will need to be able to configure the database
+    /// with only a subset of columns.  The `columns` argument specifies which
+    /// columns to configure in the database.
+    ///
+    /// Note that RocksDB is weird.  It’s not possible to open database in
+    /// read/write mode without specifying all the column families existing in
+    /// the database.  On the other hand, it’s not possible to open database in
+    /// read-only mode while specifying column families which don’t exist.
+    ///
+    /// Furthermore, note that when opening in read/write mode, we configure
+    /// RocksDB to create missing columns.
+    ///
+    /// With all that, it’s actually quite messy if at some point we’ll end up
+    /// opening cold storage as hot since it’ll create all the missing columns.
+    fn open_with_columns(
+        path: &Path,
+        store_config: &StoreConfig,
+        mode: Mode,
+        columns: impl Iterator<Item = DBCol> + Clone,
```

Comment:
I'd pass a slice -- this is a super-cold function, no need to pay monomorphisation costs here.

Reply:
I’ve also considered &dyn Iterator but I guess collecting in vector and passing by slice is indeed the simplest.